### DISCUSSION TITLE
How to create store hook to optionally accept selector and equality function with useStore

### DISCUSSION THREAD

**Discussion by jonavila** [pmndrs/zustand#1270] [Q&A] [answered]
Hi All,

I was looking at this example from v4 (https://codesandbox.io/s/polished-pond-4jn1e2?file=/src/App.tsx) and I'm wondering if anyone knows the necessary TS typing to allow `useMyStore` to optionally accept a selector function and an equality function. Currently, it takes no arguments.

**Comment by jonavila** [comment]
I found this https://github.com/pmndrs/zustand/discussions/1084#discussioncomment-3141622, but not sure how to make the selector be optional

**Comment by dai-shi** [comment, marked as answer]
It seems like to make selector optional, we need to overload function.
```ts
function useMyStore(): MyState;
function useMyStore<T>(
  selector: (state: MyState) => T,
  equalityFn?: (a: T, b: T) => boolean
): T;
function useMyStore<T>(
  selector?: (state: MyState) => T,
  equalityFn?: (a: T, b: T) => boolean
) {
  const store = useContext(MyContext);
  if (store === null) {
    throw new Error("no provider");
  }
  return useStore(store, selector as any, equalityFn);
}
```
https://codesandbox.io/s/adoring-paper-inrlu9?file=/src/App.tsx
